subsys/bluetooth/services/bas.c

```diff
@@ -0,0 +1,113 @@
+/** @file
+ *  @brief BLE GATT Battery Service
+ */
+
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <misc/byteorder.h>
+#include <misc/printk.h>
+#include <stdbool.h>
+#include <zephyr/types.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/services/bas.h>
+#include <bluetooth/uuid.h>
+
+#include <sensor.h>
+
+
+static struct bt_gatt_service bas_svc;
+static struct bt_gatt_attr attrs[];
+static struct bt_gatt_ccc_cfg blvl_ccc_cfg[BT_GATT_CCC_MAX];
+
+static struct device *battery;
+static struct sensor_value blvl = {0};
+static bool simulate_blvl;
+
+
+static void blvl_ccc_cfg_changed(const struct bt_gatt_attr *attr, u16_t value)
+{
+	simulate_blvl = (value == BT_GATT_CCC_NOTIFY) ? true : false;
+}
+
+static ssize_t read_blvl(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			 void *buf, u16_t len, u16_t offset)
+{
+	int err;
+
+	err = sensor_sample_fetch_chan(battery, SENSOR_CHAN_SOC);
+	if (err) {
+		return err;
+	}
+
+	err = sensor_channel_get(battery, SENSOR_CHAN_SOC, &blvl);
+	if (err) {
+		return err;
+	}
+
+	u8_t lvl8 = (u8_t)(blvl.val1);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &lvl8,
+				 sizeof(lvl8));
+}
+
+void bas_init(void)
+{
+	int err;
+
+	battery = device_get_binding(CONFIG_BATTERY_SENSOR_DEV_NAME);
+	if (!battery) {
+		return;
+	}
+
+	err = bt_gatt_service_register(&bas_svc);
+	if (err) {
+		return;
+	}
+}
+
+void bas_notify(void)
+{
+	int err;
+
+	if (!simulate_blvl || !battery) {
+		return;
+	}
+
+	err = sensor_sample_fetch_chan(battery, SENSOR_CHAN_SOC);
+	if (err) {
+		return;
+	}
+
+	err = sensor_channel_get(battery, SENSOR_CHAN_SOC, &blvl);
+	if (err) {
+		return;
+	}
+
+	u8_t lvl8 = (u8_t)(blvl.val1);
+
+	err = bt_gatt_notify(NULL, &attrs[1], &lvl8, sizeof(lvl8));
+	if (err) {
+		return;
+	}
+}
+
+
+/* Battery Service Declaration */
+static struct bt_gatt_attr attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_BAS),
+	BT_GATT_CHARACTERISTIC(BT_UUID_BAS_BATTERY_LEVEL,
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
+			       BT_GATT_PERM_READ, read_blvl, NULL, &blvl),
+	BT_GATT_CCC(blvl_ccc_cfg, blvl_ccc_cfg_changed),
+};
+
+static struct bt_gatt_service bas_svc = BT_GATT_SERVICE(attrs);
```

Comment:
Would it make sense to use ``SYS_INIT`` to automatically initialize the service similar to how it is done in the mgmt subsystem?

Reply:
That's probably a good idea!

Reply:
ah, you're right. I don't know whether GATT and GAP have to be first in the list of registered services.